ios/tunnel/untrusted.go

```diff
@@ -0,0 +1,483 @@
+	if err != nil {
+		return err
+	}
+	serverKey := make([]byte, 32)
+	_, err = hkdf.New(sha512.New, sessionKey, nil, []byte("ServerEncrypt-main")).Read(serverKey)
+	if err != nil {
+		return err
+	}
+	server, err := chacha20poly1305.New(serverKey)
+	if err != nil {
+		return err
+	}
+	client, err := chacha20poly1305.New(clientKey)
+	if err != nil {
+		return err
+	}
+
+	t.cipher = newCipherStream(t.controlChannel, client, server)
+
+	return nil
+}
+
+func (t *tunnelService) setupManualPairing() error {
+	buf := newTlvBuffer()
+	buf.writeByte(typeMethod, 0x00)
+	buf.writeByte(typeState, pairStateStartRequest)
+
+	event := pairingData{
+		data:            buf.bytes(),
+		kind:            "setupManualPairing",
+		startNewSession: true,
+	}
+
+	err := t.controlChannel.writeEvent(&event)
+	if err != nil {
+		return err
+	}
+	_, err = t.controlChannel.read()
+	if err != nil {
+		return err
+	}
+	return err
+}
+
+func (t *tunnelService) readDeviceKey() (publicKey []byte, salt []byte, err error) {
+	var pairingData pairingData
+	err = t.controlChannel.readEvent(&pairingData)
+	if err != nil {
+		return
+	}
+	publicKey, err = tlvReader(pairingData.data).readCoalesced(typePublicKey)
+	if err != nil {
+		return
+	}
+	salt, err = tlvReader(pairingData.data).readCoalesced(typeSalt)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (t *tunnelService) createUnlockKey() ([]byte, error) {
+	err := t.cipher.write(map[string]interface{}{
+		"request": map[string]interface{}{
+			"_0": map[string]interface{}{
+				"createRemoteUnlockKey": map[string]interface{}{},
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	var res map[string]interface{}
+	err = t.cipher.read(&res)
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, err
+}
+
+func (t *tunnelService) verifyPair() error {
+	key, _ := ecdh.X25519().GenerateKey(rand.Reader)
+	tlv := newTlvBuffer()
+	tlv.writeByte(typeState, pairStateStartRequest)
+	tlv.writeData(typePublicKey, key.PublicKey().Bytes())
+
+	event := pairingData{
+		data:            tlv.bytes(),
+		kind:            "verifyManualPairing",
+		startNewSession: true,
+	}
+
+	selfId := t.pairRecords.selfId
+
+	err := t.controlChannel.writeEvent(&event)
+
+	var devP pairingData
+	err = t.controlChannel.readEvent(&devP)
+	if err != nil {
+		return err
+	}
+
+	devicePublicKeyBytes, err := tlvReader(devP.data).readCoalesced(typePublicKey)
+	if err != nil {
+		return err
+	}
+
+	devicePublicKey, err := ecdh.X25519().NewPublicKey(devicePublicKeyBytes)
+	if err != nil {
+		return err
+	}
+
+	sharedSecret, err := key.ECDH(devicePublicKey)
+	if err != nil {
+		return err
+	}
+
+	derived := make([]byte, 32)
+	_, err = hkdf.New(sha512.New, sharedSecret, []byte("Pair-Verify-Encrypt-Salt"), []byte("Pair-Verify-Encrypt-Info")).Read(derived)
+	if err != nil {
+		return err
+	}
+
+	ci, err := chacha20poly1305.New(derived)
+	if err != nil {
+		return err
+	}
+
+	signBuf := bytes.NewBuffer(nil)
+	signBuf.Write(key.PublicKey().Bytes())
```

Comment:
you are swallowing a few errors here I think. Since Golang 1.20 it finally supports multierrs out of the box. 
`err := errors.Join(err1, err2)`
You can join them like that if it does not make sense to handle them all individiually.

Reply:
`Write` on a `bytes.Buffer` never returns an error. I'm pretty sure it's handled in other places, but it's actually not necessary

Reply:
```Write appends the contents of p to the buffer, growing the buffer as needed. The return value n is the length of p; err is always nil. If the buffer becomes too large, Write will panic with [ErrTooLarge](https://pkg.go.dev/bytes#ErrTooLarge).```
oh man.. eyebleed. But ok, then. I did not know this. Maybe nice to add a comment to explain this. It is quite unusualy to not handle errors.